### PR TITLE
Document and test ListView.showOnEmpty, see #11627

### PR DESCRIPTION
**Pull request by mytskine**
This addresses #11627 by completing the documentation without changing the behavior. Some basic unit tests are added on ListView to ensure this won't change unintentionnally.

That means a ListView widget with an empty provider and `showOnEmpty=true` will still display an empty element (by default, an empty "div.list-view").

### CODE REVIEW

**Comment by dynasource** [comment]
Thanks for putting effort in creating unit tests. Would you be willing to put the tests in smaller units to prove your usecase?

**Comment by mytskine** [comment]
I'm sorry but I don't understand your answer. What do you mean with "to
prove your usecase"? What I want to "prove" is that the documentation of
this property was bad, as can be seen in #11627.

Does "put the tests in smaller units" means that tests should have only
one assertion?

I only put assertions on the output controlled by the `showOnEmpty`
property, so I can't see any smart way to split those units.  If that's
your wish, I'll remove the non-essential assertions. I'd keep
"!showOnEmpty displays the empty message" and "showOnEmpty contains no
text".

**Comment by dynasource** [comment]
if its your goal to have these unit tests in core (to prove/cover your usecase), you should further cut these in blocks like doing assertings on subfunctions instead of the run() function

**Comment by mytskine** [comment]
That makes no sense because the property I want to test is only used in
`ListView.run()` — more exactly, in the abstract parent class
[BaseListView](https://github.com/mytskine/yii2/blob/master/framework/widgets/BaseListView.php#L124)

**Comment by dynasource** [comment]
you're right, didnt know run() was that bad. Perhabs your can also update ListView::run and move the logic for setting $content to its own getter (protected function getContent)

Then its easier to test your usecase in a more contained enviroment.

**Comment by SilverFire** [comment]
@mytskine the tests are failing now. Is it expected?

**Comment by mytskine** [comment]
These errors already occured before I changed anything. I thought it was
because of my environment. Moreover, the error messages involve
ActiveRecordTestTrait which is out of my concern.

The widgets tests, where I added on test case, do not fail.

The official yii2/master has the same failing tests:
<https://travis-ci.org/yiisoft/yii2>

**Comment by mytskine** [comment]
The unrelated unit tests are now fixed on master, see #12745. I've rebased the branch.

File: tests/framework/widgets/ListViewTest.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        ob_start();
+        echo ListView::widget([
+            'dataProvider' => $dataProvider,
+            'showOnEmpty' => false,
+            'emptyText' => "Nothing at all",
+        ]);
+        $actualHtml = ob_get_clean();
+
+        $this->assertTrue(strpos($actualHtml, "Nothing at all") !== false, "displays the empty message");
```
Comment (dynasource) [inline tests/framework/widgets/ListViewTest.php]: these assertions are a but difficult to read. Can you use `assertContains` or `assertNotContains`

File: framework/widgets/BaseListView.php [hunk 2]
```diff
@@ -69,7 +69,9 @@
      */
     public $summaryOptions = ['class' => 'summary'];
     /**
-     * @var boolean whether to show the list view if [[dataProvider]] returns no data.
+     * @var boolean whether to show an empty list view if [[dataProvider]] returns no data.
+     * The default value is false which displays an element according to the `emptyText`
+     * and `emptyTextOptions` properties.
```
Comment (dynasource) [inline framework/widgets/BaseListView.php]: I agree with that this is confusion. With a `false` for showOnEmpty something is actually "shown". Can we fix this problem in this PR  by renaming the $property (for 2.1 and future releases)

**Review by dynasource — CHANGES_REQUESTED** [review]
thanks for your efforts. Note: There are some conflicts with the master branch.

**Review by samdark — APPROVED** [review]

**Review by cebe — APPROVED** [review]

**Comment by SilverFire** [comment]
Merged with simplified tests (https://github.com/yiisoft/yii2/commit/0261a5498f1fbbe9e84dc181c6fc8cada94f25bd).

Thank you!
